### ISSUE TITLE
Identifying the line number of an error

### ISSUE THREAD

**Issue by oTree-org** [dmulholl/ibis#1]
Hi, I just found this project and really like it! Clean code and good API design 👍 I'm thinking of using it in my project oTree: https://github.com/oTree-org/oTree

It would be really useful if exceptions reported which line in the template the error occurred on, so people could debug more easily. Currently this is what I get:

```
File ".\ibis_test.py", line 5, in <module>
    template = ibis_loader('tpl.html')
  File "C:\project\lib\site-packages\ibis\loaders.py", line 50, in __call__
    return self.cache.setdefault(path, Template(file.read()))
  File "C:\project\lib\site-packages\ibis\templates.py", line 24, in __init__
    self.root = self._parse(Lexer(string))
  File "C:\project\lib\site-packages\ibis\templates.py", line 54, in _parse
    node = nodes.nodemap[token.tag](token)
  File "C:\project\lib\site-packages\ibis\nodes.py", line 155, in __init__
    self.process_token(token)
  File "C:\project\lib\site-packages\ibis\nodes.py", line 394, in process_token
    raise TemplateSyntaxError(msg) from None
ibis.errors.TemplateSyntaxError: malformed [if] tag: [if]
```

Having the file name and line number as attributes on the exception object (e.g. `.lineno`) would help authors of other tools, e.g. to show an error page that displays the error in context. However, it seems Ibis does not currently track line numbers internally.

What do you think?

**Comment by oTree-org** [comment]
In jinja, the location of the error in the template is actually a stack frame:

```
  File "C:\project\lib\site-packages\jinja2\environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "C:\project\lib\site-packages\jinja2\_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File ".\tpl.html", line 8, in template
    {% if x %}
jinja2.exceptions.TemplateSyntaxError: Unexpected end of template. Jinja was looking for the following tags: 'elif' or 'else' or 'endif'. The innermost block that needs to be closed is 'if'.
```

**Comment by dmulholl** [comment]
Glad to hear you're finding it useful! You're absolutely right, the vague error messages are the library's weakest feature, and an issue I've been meaning to get around to fixing for a long time. 

Your timing is excellent, I actually wrote up a [tutorial](http://www.dmulholl.com/lets-build/a-template-language.html) on building a template language a couple of weeks back which has left me with an itch to sort out Ibis's error reporting. I have a plan for it that I should have in place by the weekend.

(I really like Jinja's error message by the way - I think I'm going to have to steal that format!)

All that said, you should seriously consider using Jinja for your project if it's an option - it's by far the most optimised and battle-tested template language available for Python. Ibis is something of an accident that happened because I was interested in how template engines worked, built one to find out, then got carried away and kept adding features until it could do everything I needed...

**Comment by oTree-org** [comment]
> Your timing is excellent, I actually wrote up a [tutorial](http://www.dmulholl.com/lets-build/a-template-language.html) on building a template language a couple of weeks back which has left me with an itch to sort out Ibis's error reporting. I have a plan for it that I should have in place by the weekend.

That's great to hear! Let me know as you progress and I'll be sure to devote time to testing & feedback. You can also contact me at chris@otree.org.

> 
> (I really like Jinja's error message by the way - I think I'm going to have to steal that format!)

Yes it seems ideal because it integrates naturally with web frameworks' debug-error pages that show each stack frame in context. For example here is Starlette:

![image](https://user-images.githubusercontent.com/5423929/100834245-883f5700-34a6-11eb-9827-8f3759b530d2.png)

> 
> All that said, you should seriously consider using Jinja for your project if it's an option - it's by far the most optimised and battle-tested template language available for Python. Ibis is something of an accident that happened because I was interested in how template engines worked, built one to find out, then got carried away and kept adding features until it could do everything I needed...

Yes I am testing out Jinja also, but would like to give Ibis a chance because:

1. It meshes better with my project's design (simple API, accessible to newcomers)
2. I prefer to keep my dependencies small
3. Once I switch to Jinja, there is no turning back, since it has lots of tags & constructs not found in Ibis/Django, so it's very hard for my users to migrate their projects off of it. Rather, I can try Ibis, and if it doesn't work, we can turn to Jinja, which is kind of like a superset.

**Comment by dmulholl** [comment]
I get that reasoning, I always like using the simplest tool possible myself.

I just pushed an update on `master`, version `2.0.0-alpha.1`, also available from the package index as

    pip install ibis==2.0.0a1

Internally it's a fairly substantial refactor to improve error handling across the board, but the external API hasn't changed very much:

* The `Template` class now takes an optional `template_id` argument for use in error messages.
* The exception hierarchy is significantly simpler. 
* Exception types now include template IDs and line numbers where possible.
* The `config` module is gone - you now specify a template loader directly on the `ibis` module.

The new v2 docs are available here: http://www.dmulholl.com/docs/ibis/master/

The old v1 docs are still available here: http://www.dmulholl.com/docs/ibis/v1/

I'm sure there'll be bugs so if you get the opportunity to test it and find any, I'd love to hear about them!

**Comment by oTree-org** [comment]
Great! I'm taking a look now!

**Comment by dmulholl** [comment]
Thanks for all the feedback, it was really helpful. I'm planning to finalise the 2.0 release next weekend, I'll bump this thread when I do. After that Ibis should be stable for the long term. v1 went five years without a breaking change, I'm sure v2 can match that record!

**Comment by oTree-org** [comment]
Hi, it looks much better! And I am productively using it now 👍 

Although TemplateError is raised in many places now, I think it would be good to do this raising at a more outer level rather than inside each template tag. Currently, some errors can still slip through without reporting a line number. For example:

```
t = Template("""
{% include a %}
""")
res = t.render({'a': None})
print(res)
```

Gives:

```
Traceback (most recent call last):
  File ".\ibis_test.py", line 24, in <module>
    res = t.render({'a': None})
  File "c:\project\ibis\template.py", line 28, in render
    return self.root_node.render(context.Context(data_dict, self, strict_mode))
  File "c:\project\ibis\nodes.py", line 157, in render
    return ''.join(child.render(context) for child in self.children)
  File "c:\project\ibis\nodes.py", line 157, in <genexpr>
    return ''.join(child.render(context) for child in self.children)
  File "c:\project\ibis\nodes.py", line 507, in render
    template = ibis.loader(template_name)
  File "c:\project\ibis\loaders.py", line 33, in __call__
    path = os.path.join(self.root_dir, filename)
  File "C:\Users\wicke\AppData\Local\Programs\Python\Python37-32\lib\ntpath.py", line 115, in join
    genericpath._check_arg_types('join', path, *paths)
  File "C:\Users\wicke\AppData\Local\Programs\Python\Python37-32\lib\genericpath.py", line 149, in _check_arg_types
    (funcname, s.__class__.__name__)) from None
TypeError: join() argument must be str or bytes, not 'NoneType'
```
How about on the Node base class, we wrap the calls to .process_token() and .render()?

    def outer_process_token(self, token):
        try:
            self.process_token(token)
        except TemplateError:
            # this will already have all the info we need
            raise
        except Exception as exc:
            raise TemplateSyntaxError(
                f'{token.keyword}: {exc}', token.template_id, token.line_number
            ) from exc

    def outer_render(self, context):
        token = self.token
        try:
            return self.render(context)
        except TemplateError:
            raise
        except Exception as exc:
            raise TemplateRenderingError(
                f'{token.keyword}: {exc}', token.template_id, token.line_number
            ) from exc

This way:
- if there is any uncaught exception inside a template tag, we will know the line number where it happened.
- Using the "raise from" and showing the original exception message ensures that we do not lose any info about the original exception.
- the "except TemplateError" is so that if a TemplateSyntaxError/TemplateRenderingError/etc was raised inside a template tag, it already has the info we need (line number etc), so no need to change it.
- Using this approach reduces the boilerplate needed inside each template tag. For example, repeating the message about template name, tag name, line number.

I see that errors in template filters are handled at an outer level (`_apply_filters_to_variable`) rather than inside the implementation of each filter.

By the way how about we pass token.keyword as a 3rd attribute to TemplateError?

```
class TemplateError(Exception):

    def __init__(self, msg, template_id=None, line_number=None, tag_name=None):
        super().__init__(msg)
        self.template_id = template_id
        self.line_number = line_number
        self.tag_name = tag_name
```

**Comment by dmulholl** [comment]
Nice catch - you have a real knack for finding these edge cases! Turns out the `{% include %}` tag needed a lot more error checking, that's fixed in `2.0.0-alpha.7`. 

I like your idea for wrapping unexpected errors. I've implemented it for `.process_token()` as I could do it cleanly in the `Node` constructor and any exceptions we catch are guaranteed to be template-specific syntax errors. 

I'm going to hold off on implementing the rendering-error equivalent until I've had a chance to do some more experimenting as I'm wary of intercepting application-specific exceptions in an unexpected way. I do like the idea though.

I've also reorganised the exception classes to include the `Token` instance where possible. That automatically gives access to the template_id, line_number, keyword, etc., along with any future attributes that might get added.

**Comment by oTree-org** [comment]
Great thank you! Looking forward to trying it out!

**Comment by dmulholl** [comment]
Almost done! Barring unforeseen disaster I'm planning to release the current version, `2.0.0-rc1`, tomorrow as `2.0`.

Don't worry, I'll still be open to feedback, and if some design decision turns out to have been a horrible mistake and I have to cut a v3 release at some point to fix it, it's not the end of the world.

Thanks for all your feedback so far, it's been really helpful!

**Comment by oTree-org** [comment]
OK thank you i will take a look!  by the way, could you send me an email at chris@otree.org? So that I have your email address, I want to send some snippets & things that are not bug reports but worth discussing.

**Comment by dmulholl** [comment]
I'm going to close this issue as I've just released v2.0. Thanks for all the feedback, feel free to open a new issue if you find any lurking bugs!
